ERCS/erc-7699.md

```diff
@@ -0,0 +1,212 @@
+
+Currently, users and merchants who want to include a payment reference in their transactions must rely on off chain external systems or custom payment proxy implementations. In traditional finance systems, payment references are often included in wire transfers and other types of electronic payments, making it easy for users and merchants to manage and reconcile their transactions. Such as:
+
+ - SWIFT MT103: field 70 “Remittance Information” is commonly used for such content (e.g " PAYMENT FOR INVOICE 998877"). There is also field 72 “Sender to receiver information”.
+ - ISO 20022 (for SEPA): PAIN.001 has field called RmtInf (Remittance Information)
+
+By extending the existing ERC-20 token standard with payment transfer reference capabilities, this proposal will help bridge the gap between traditional finance systems and the world of decentralized finance, providing a more seamless experience for users, merchants, and service providers alike.
+
+## Specification
+
+The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.
+
+Any contract complying with ERC-20 when extended with this ERC, MUST implement the following interface:
+```
+// The EIP-165 identifier of this interface is 0x1522573a
+
+interface IERC7699 {
+
+function transfer(address to, uint256 amount, bytes calldata transferReference) external returns (bool);
+
+function transferFrom(address from, address to, uint256 amount, bytes calldata transferReference) external returns (bool);
+
+event TransferReference(bytes32 indexed loggedReference);
+
+}
+```
+
+These `transfer` and `transferFrom` functions, in addition to the standard transfer behaviour, MUST emit a `transferReference` event with a `loggedReference` parameter (with only exception defined below).
+
+The corresponding ERC-20 `Transfer` event MUST be emitted following the `TransferReference` event, ideally immediately afterwards for the client to be able to seek the associated `Transfer` event log record. 
+
+Emitted `loggedReference` MIGHT be the exact copy of the `transferReference` (when less then 33 bytes) or the derived data from the rich `transferReference` structure and other processing. This is up to the implementer. One MUST NOT expect the `transferReference` and `loggedReference` data are equal.
+
+The `loggedReference` parameter MAY contain any data of bytes32 type.
+
+The `transferReference` parameter MAY be empty. In this case and only in this case the `TransferReference` event MUST NOT be emitted, effectively mimicking the regular ERC-20 transfer without any transfer reference. 
+
+The `transferReference` parameter is not limited in length by design, users are motivated to keep it short due to calldata and execution gas costs.
+
+The `TransferReference` event MUST NOT be anonymous. This is to ensure that the event signature is logged and can be used as a filter.
```

Comment:
This _might_ be clear to someone reading this slower than I am, but I feel "anonymous" here is a bit vague. Might be something to fix while in Draft.

Reply:
ah, it refers to the type of event, resp. to the specifier as documented here: https://docs.soliditylang.org/en/latest/contracts.html#events

Would this be better?
"""
The `TransferReference` event MUST NOT be declared with the `anonymous` specifier.
"""

Reply:
fixed with #716  Thx